### ISSUE TITLE
MsgpackSerializer function not defined

### ISSUE THREAD

**Issue by MKirchen** [crossbario/autobahn-js#236]
I cannot get msgpack serialization to work in the browser. It seems like this part in serializer.js is never called:

``` javascript
try { ...
   function MsgpackSerializer() {
      this.SERIALIZER_ID = 'msgpack';
      this.BINARY = true;
      this.codec = msgpack.createCodec();
   }
}
```

Minimal test example:

``` html
<html>
   <body>
      <h1>MsgpackSerializer?</h1>
      <script>AUTOBAHN_DEBUG = true;</script>
      <script src="https://autobahn.s3.amazonaws.com/autobahnjs/latest/autobahn.js"></script>
      <script>
         console.log(autobahn.serializer.MsgpackSerializer)
         console.log(autobahn.serializer.JSONSerializer)
      </script>
   </body>
</html>
```

prints the following in the console:

``` javascript
null
function JSONSerialize()
```

I am very unexperienced with javascript and could not fix it by myself so far.

**Comment by MKirchen** [comment]
I probably found the reason why MsgpackSerializer is null. In serializer.js these two lines (35-36):
```javascript
   var write_type = require("msgpack-lite/lib/write-type").type;
   var write_token = require("msgpack-lite/lib/write-token").token;
```
will work fine with msgpack-lite@0.1.17 but not with the newest version msgpack-lite@0.1.26. When changing the autobahn dependecies in package.json from "msgpack-lite": ">= 0.1.17" to "msgpack-lite": "= 0.1.17" and installing with npm everything works fine and MsgpackSerializer is correctly initialized.

In version 0.1.26 of msgpack-lite in the file write-type.js there is a function getWriteType now:
```javascript
exports.getWriteType = getWriteType;

function getWriteType(options) {
  var token = WriteToken.getWriteToken(options);
  var useraw = options && options.useraw;
  var binarraybuffer = HAS_UINT8ARRAY && options && options.binarraybuffer;
  var isBuffer = binarraybuffer ? Bufferish.isArrayBuffer : Bufferish.isBuffer;
  var bin = binarraybuffer ? bin_arraybuffer : bin_buffer;
  var usemap = HAS_MAP && options && options.usemap;
  var map = usemap ? map_to_map : obj_to_map;

  var writeType = {
    "boolean": bool,
    "function": nil,
    "number": number,
    "object": (useraw ? object_raw : object),
    "string": _string(useraw ? raw_head_size : str_head_size),
    "symbol": nil,
    "undefined": nil
  };

  return writeType;
```

In version 0.1.17 of msgpack-lite in the file write-type.js this was different:
```javascript
exports.type = {
  "boolean": bool,
  "function": nil,
  "number": number,
  "object": object,
  "string": string,
  "symbol": nil,
  "undefined": nil
};
```

I think adapting the code in serializer.js to:
```javascript
   var write_type = require("msgpack-lite/lib/write-type").getWritetype();
   var write_token = require("msgpack-lite/lib/write-token").getWritetoken();
```
solves this issue for newer versions of msgpack-lite. In my test, MsgpackSerializer is correctly initialized then. However (as I am very new to javascript & autobahn in general) I dont know if anything else is maybe broken then.

**Comment by oberstet** [comment]
Oh, I see, the msgpack change needs a compatibility fix in Autobahn. Thanks for tracking this down! 

I think best would be to add code in AutobahnJS that detects the presence of `getWriteType` in msgpack, and if present, use the code you suggest, otherwise fallback to the existing behavior. This way, it should support both older and new msgpack versions.

**Comment by MKirchen** [comment]
I managed to browserify the adapted version of autobahn-js (including getWriteType() and getWriteToken()) and my minimal test example works now.

```html
<html>
   <body>
      <h1>MsgpackSerializer?</h1>
      <script>AUTOBAHN_DEBUG = true;</script>
      <script src="https://rawgit.com/kawanet/msgpack-lite/master/dist/msgpack.min.js"></script>
      <script src="autobahn.js"></script>
      <script>
         console.log(autobahn.serializer.MsgpackSerializer)
         console.log(autobahn.serializer.JSONSerializer)
      </script>
   </body>
</html>
```

However, I need to additionally include:
```html
      <script src="https://rawgit.com/kawanet/msgpack-lite/master/dist/msgpack.min.js"></script>
```
This was not clear to me before, as I thought browserify will explicitly include all dependecies (sorry, I am very new to all of this). However, I assume this is because in package.json it says:
    "browser": {
        "lib/transport/rawsocket.js": false,
        "msgpack-lite": "**msgpack-lite/global**"
    }
and because of this the source code of msgpack-lite is not included and one needs to additionally add this in the .html. Is that correct? (see also https://github.com/kawanet/msgpack-lite#messagepack-with-browserify)

I attached the browserified version that should work now with msgpack-lite@0.1.26
[autobahn.js.zip](https://github.com/crossbario/autobahn-js/files/531374/autobahn.js.zip)

**Comment by MKirchen** [comment]
My initial motivation was to send raw binary data with msgpack from a python client -> crossbar -> javascript client in the browser (In the crossbar config.json, I therefore only allowed msgpack as serializer).

When I tried this first, the MsgpackSerializer was not recognized properly but as discussed above, this is fixed now (at least there is a way to fix it - please note that the official build https://autobahn.s3.amazonaws.com/autobahnjs/latest/autobahn.js is still broken [even when using msgpack-lite@0.1.17] as it was apparently build using a newer version of msgpack-lite without taking care of the compatibility issues discussed above [e.g. the file write-type.js of the new msgpack version is included in the browserified version of autobahn.js, while in serializer.js autobahn expects to have the old version of write-type.js available]).

However, once msgpack is available, new problems occur. For example, it is not possible to unserialize the message that comes from the python client using msgpack-lite, as it is send as ArrayBuffer and msgpack-lite can only unserialize it if its an Uint8Array (see https://github.com/kawanet/msgpack-lite/issues/44), which is strange because in their Readme.md they claim that they can decode ArrayBuffers.

Thus, (in order to just decode the data) I had to add the following line to websocket.js
```javascript
         websocket.onmessage = function (evt) {
            log.debug("WebSocket transport receive", evt.data);
            var raw_binary_data = new Uint8Array(evt.data); // needed to transform to Uint8Array first...
            var msg = transport.serializer.unserialize(raw_binary_data);
            console.log("onmessage", msg)
            transport.onmessage(msg);
         }
```
which is probably not a good idea...

Still, I was then finally able to decode this message, but as a consequence this then caused bugs with the Autobahn Python unserializer (serializer.py), when receiving messages from the javascript client websocket. Somehow, the raw message data arriving at crossbar contained a float and Klass.parse(msg) gave a protocol error. I dont really know exactly what happened there.

My question is now: Do you have a working example to send raw binary data from a python client to a javascript browser client using msgpack?

Moreover: Maybe it would also be a good idea to add a binary data type to the msgpack test of Autbahn JS.

Btw: Sending binary data from a python client to another python client with msgpack works perfectly fine. So, it seems like these bugs are really Autobahn JS related.

**Comment by MKirchen** [comment]
I think I solved this. Will try to do a PR or write solution tomorrow.

**Comment by oberstet** [comment]
fixed in v0.11.0
